cudax/include/cuda/experimental/__event/event.cuh

```diff
@@ -0,0 +1,141 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of CUDA Experimental in CUDA C++ Core Libraries,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _CUDAX_EVENT_DETAIL_H
+#define _CUDAX_EVENT_DETAIL_H
+
+#include <cuda_runtime_api.h>
+// cuda_runtime_api needs to come first
+
+#include <cuda/std/detail/__config>
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+#include <cuda/std/__cuda/api_wrapper.h>
+#include <cuda/std/__exception/cuda_error.h>
+#include <cuda/std/cstddef>
+#include <cuda/std/utility>
+
+// CUDAX headers here
+#include <cuda/experimental/__detail/utility.cuh>
+#include <cuda/experimental/__event/event_ref.cuh>
+
+namespace cuda::experimental
+{
+class timed_event;
+
+//! @brief An owning wrapper for an untimed `cudaEvent_t`.
+class event : public event_ref
+{
+  friend class timed_event;
+
+public:
+  //! @brief Flags to use when creating the event.
+  enum class flags : unsigned int
+  {
+    none          = cudaEventDefault,
+    blocking_sync = cudaEventBlockingSync,
+    interprocess  = cudaEventInterprocess
+  };
+
+  //! @brief Construct a new `event` object with timing disabled.
+  //!
+  //! @throws cuda_error if the event creation fails.
+  event()
+      : event(static_cast<unsigned int>(cudaEventDisableTiming))
+  {}
```

Comment:
i'm having misgivings about this. generally, types that have a moved-from state also have default constructors that leave the object in the moved-from state. i want to change this to default-construct into the null state and add an `event::create` static function for creating a new CUDA event.

Reply:
Would an alternative be creation on first record? This would also allow us to not take the device explicitly and instead deduce it from what we are recording it into

Reply:
:thinking: then the `event` and `timed_event` types would need to store the flags from the constructor. :bulb: but we can use the lower bits of the `__event_` data member for that.

thanks, i'll try to implement that. might also satisfy cedric, who wants to create an event from a stream like `stream.create_event()`.